From: Alan Shearer <alan@localhost>
Date: Tue, 22 Sep 2026 04:38:00 -0500
Subject: virtio_block: ignore zero blk_size

set_capacity divides capacity by config.blk_size when
VIRTIO_BLK_F_BLK_SIZE is negotiated. A host that advertises
the feature with blk_size == 0 divides by zero and KDLs,
including from the config-change interrupt.

Keep the 512-byte default when the advertised size is zero.

---
 src/add-ons/kernel/drivers/disk/virtual/virtio_block/virtio_block.cpp | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/src/add-ons/kernel/drivers/disk/virtual/virtio_block/virtio_block.cpp b/src/add-ons/kernel/drivers/disk/virtual/virtio_block/virtio_block.cpp
index 9eb71254daec..c2c0d0d0d0d0 100644
--- a/src/add-ons/kernel/drivers/disk/virtual/virtio_block/virtio_block.cpp
+++ b/src/add-ons/kernel/drivers/disk/virtual/virtio_block/virtio_block.cpp
@@ -444,8 +444,9 @@ virtio_block_set_capacity(virtio_block_driver_info* info)
 {
 	// get capacity
 	uint32 blockSize = 512;
-	if ((info->features & VIRTIO_BLK_F_BLK_SIZE) != 0)
+	if ((info->features & VIRTIO_BLK_F_BLK_SIZE) != 0
+		&& info->config.blk_size != 0)
 		blockSize = info->config.blk_size;
 	uint64 capacity = info->config.capacity * 512 / blockSize;
 	uint32 physicalBlockSize = blockSize;
